### DISCUSSION TITLE
Create profile with extra data from authentication

### DISCUSSION THREAD

**Discussion by orisa** [supabase/supabase#9049] [Questions] [answered]
Still new to Supabase but I'm trying this use case: 
   on signup, 
       create a profile table with extra data in the signup request's payload
Given:
   `  payload: 
         {birthday: "some date", business_title: "some title", email: "some email" password" }`
Is it possible to get the birthday and business_title to add a row to the profile table. Currently, I've got 2 solution ideas:
1 - Function triggers: 
     Todo: on insert to users table, add a row to profile table. 
     Issue: I can get email. how do I get birthday and business title
2-  Subabase function hooks:
      Todo:  write a server less function that makes two requests. one for signup with just email and password.               Another adding a row to profile table if signup was successful.
       Issue: What will I do if adding a profile fails but signup was successful. Will I delete the the user. Or I     would have to account for this scenario in the client app

**Comment by bennik88** [comment, marked as answer]
1. To your first question. There are already a lot of discussions about this topic. I hope this will help you out. Function triggers are a great way to do this. The way to access the `raw_user_meta_data` from your sign-up request would look like this:
**Function**
```sql
begin
  insert into public.profiles(id, email, birthday, business_title)
  values(new.id, new.email, new.raw_user_meta_data->>'birthday', new.raw_user_meta_data->>'business_title');

  return new;
end;
```
Here are some additional resources for functions:
[Create PostgreSQL Functions with Supabase - Supabase Youtube](https://www.youtube.com/watch?v=MJZCCpCYEqk)
[Handle New User - Function & Trigger](https://database.dev/handle-new-user)
[How to add additional metadata to users table? #6363](https://github.com/supabase/supabase/discussions/6363)

The approach with function and triggers will also sort you out when it comes to a sign-up that was unsuccessful as it wouldn't INSERT the entry in your table.

**Reply by orisa** [reply]
The code runs but birthday and business_title are NULL. When I make the signup request using postman the raw_user_meta_data was not present. Does Supabase create that on the backend? How do I query raw_user_meta_data of users from the supabase console?  Pls note I'm using bash/http request api for my sup abase requests. I'm new to postgresql.

**Reply by bennik88** [reply]
That is not a feature of Supabase but more a feature of PostgreSQL. Did you set up the trigger as well? As shown in my additional resourced links? You need the trigger that will be triggered when an `auth.user` will be inserted and a function that the trigger will call. When you're not creating the trigger the function will be there but never called.
```sql
create trigger on_auth_user_created
  after insert on auth.users
  for each row execute procedure public.handle_new_user();
```
Here is another resource for triggers:
[Using PostgreSQL triggers to automate processes with Supabase](https://www.youtube.com/watch?v=0N6M5BBe9AE&t=2s)
[PostgreSQL Docs CREATE TRIGGER](https://www.postgresql.org/docs/current/sql-createtrigger.html)

I suggest you watch a few PostgresSQL basic videos this will help you understand `SELECTS`and so on. I myself realized that you have to know PostgreSQL or SQL to use Supabase. So this will help you a lot and you will have a lot of great realizations when looking at their UI ;)
Here is how you query the auth users.
```sql
SELECT * FROM auth.users;
```

Selecting JSON from the `raw_user_meta_data` is also something you find with one google search which lines up with my recent answer which used the `->>` operator.
[PostgreSQL JSON](https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-json/)
All the best.

**Reply by orisa** [reply]
Thanks. I've tried most of those resources  but I didn't see how to pass additional data during signup using bash.But I've just found the solution . The problem needs to be redefined to  "how to pass additional data on signup".
For anyone who's using bash this should be the payload structure(where value of data is extra data):
```
{
    "email": "some email",
    "password": "some password",
    "data": {
         "birthdate": "some date",
        "business_title": "some_title"
    }
}
```
Then, define a supabase function as above by @bennik88 along with its trigger. 
For client side library, checkout:  [https://stackoverflow.com/questions/73185811/add-aditional-user-information-when-signup-supabase](https://stackoverflow.com/questions/73185811/add-aditional-user-information-when-signup-supabase)
